### PR TITLE
Displaying entities on terrain

### PR DESCRIPTION
**Pull request by tfili**
@mramato Can you take a look at the implementation for the RectangleGeometry? Once we get the correct way we want to do this, I'll hook it up for the other types.

- [x] Corridor
- [x] Ellipse
- [x] Polygons
- [x] Rectangle
- [x] Billboards

Added by @mramato
- [ ] Camera `flyTo` and `zoomTo` of an entity on terrain.
- [ ] Camera tracking of an entity on terrain
- [ ] Selection indicator of an entity on terrain
- [x] Detecting if `EXT_FRAG_DEPTH` is missing and fallback to non-ground visualization; otherwise this branch breaks IE, Safari, and various mobile devices.
- [ ] Point and Model clamping
- [ ] Performance

### CODE REVIEW

**Comment by mramato** [comment]
FYI, there is already an `entity-clamp-to-ground` branch for billboards and labels, but it needs work to be finished.  I would recommend eventually merging your efforts with that branch.

From a purely functional standpoint, here are the major problems I see that prevent us from having official Entity support for terrain.

__General__
1. There are problems with both the camera tracking and selection indicator because it is positioned where it thinks the entity is, not where it is actually drawn. Essentially, we need expose a way to easily query where an object is beind rendered, which shouldn't be too hard but will probably require some primitive tweaks and exposing some new functions.  This will get tricky pretty quick because of the way camera/selection tracking currently work.  It's a totally solvable problem though.

__Geometry__
1. The GroundPrimitive can't currently be batched, so performance is terrible when there are a bunch of them on the screen. Just 900 rectangles drops my fps to 20 (code below).
2. Geometry outlines can't be clamped to ground which will confuse users who will end up reporting it as a bug.
3. Geometry materials other than color doesn't work, which will also confuse users who will end up reporting it as a bug (but may not be as big of a deal as outlines).
4. Obviously lines aren't supported yet. I don't think using corridors as lines will work well at all, but I'm willing to entertain the idea and see where it ends up. (and they don't support materials/outline either)
5. I also noticed that if we go with the "no height/no extrusion means on terrain" approach, it can potentially break existing apps (who would need to set height to 0 explicitly as a workaround).  One solution would be to not put the geometry on terrain if it has a non-color material instead of just drawing it white (which is what you're doing now).

__Billboards, Labels, Points, and Models__
1. Neither PointPrimitive or Model support clamping to ground yet, but these are both often requested features that we should address.
2. No lines on terrain means that having a moving object clamped to ground will cause the line to not be clamped to ground, producing weird visual inconsistency which will confuse users who will definitely end up reporting it as a bug.

If we try and introduce terrain support to the Entity API without addressing the above issues, I think support will be very inconsistent and frustrate/confuse a lot of our users.  I'm probably willing to take incremental steps, but we need a solid plan going forward to that things don't languish.

Here's the test code I used to look into rectangle performance I mentioned above:

```javascript
var viewer = new Cesium.Viewer('cesiumContainer');
viewer.scene.debugShowFramesPerSecond = true;

Cesium.Math.setRandomNumberSeed(0);
var granularity = 3;
var i = 0;
for (var lon = -45.0; lon < 45.0; lon += granularity) {
    for (var lat = -45.0; lat < 45.0; lat += granularity) {
        i++;

        viewer.entities.add({
            rectangle : {
                coordinates : Cesium.Rectangle.fromDegrees(lon, lat, lon + granularity, lat + granularity),
                material : Cesium.Color.fromRandom({
                    alpha : 0.75
                })
            }
        });
    }
}
console.log(i);
```

File: Source/DataSources/dynamicGeometryGetBoundingSphere.js [hunk 1]
```diff
@@ -25,12 +29,14 @@ define([
...
 
         var attributes;
+        var modelMatrix;
 
         //Outline and Fill geometries have the same bounding sphere, so just use whichever one is defined and ready
         if (defined(primitive) && primitive.show && primitive.ready) {
             attributes = primitive.getGeometryInstanceAttributes(entity);
             if (defined(attributes) && defined(attributes.boundingSphere)) {
-                BoundingSphere.transform(attributes.boundingSphere, primitive.modelMatrix, result);
+                modelMatrix = defaultValue(primitive.modelMatrix, Matrix4.IDENTITY);
```
Comment (mramato) [inline Source/DataSources/dynamicGeometryGetBoundingSphere.js]: Why is this needed?  An undefined `modelMatrix` sounds like a bug to me.

File: Source/DataSources/RectangleGeometryUpdater.js [hunk 2]
```diff
@@ -556,24 +582,46 @@ define([
...
+                    geometryInstance : new GeometryInstance({
+                        id : entity,
+                        geometry : new RectangleGeometry(options),
+                        attributes: {
+                            color: ColorGeometryInstanceAttribute.fromColor(currentColor)
+                        }
+                    }),
+                    asynchronous : false
+                }));
+            }
```
Comment (mramato) [inline Source/DataSources/RectangleGeometryUpdater.js]: Formatting, else goes on the same line as the bracket.

File: Source/DataSources/RectangleGeometryUpdater.js [hunk 3]
```diff
@@ -527,16 +541,28 @@ define([
...
+        }
         primitives.removeAndDestroy(this._outlinePrimitive);
         this._primitive = undefined;
         this._outlinePrimitive = undefined;
 
-        var geometryUpdater = this._geometryUpdater;
         var entity = geometryUpdater._entity;
         var rectangle = entity.rectangle;
-        if (!entity.isShowing || !entity.isAvailable(time) || !Property.getValueOrDefault(rectangle.show, time, true)) {
+        var isAvailable = entity.isAvailable(time);
```
Comment (mramato) [inline Source/DataSources/RectangleGeometryUpdater.js]: Why was this line changed, `isAvailable` doesn't appear to be used anywhere.

**Comment by mramato** [comment]
In general, this approach doesn't look too bad, and may be okay for a short-term fix (as far as exposing current primitive capabilities).  Longer term we really need to invest time in reworking visualization to better handle the new features and additional use cases that have come about in the past year.

We probably want to handle GeoJSON in a separate pull request (as well as improve KML support in another).  CZML shouldn't need any changes.

I mentioned this above, but I think the disconnect in geometry features is going to be really confusing, we need to figure out a way to address that or it will be a usability nightmare.

Add unit tests.

File: Source/DataSources/dynamicGeometryGetBoundingSphere.js (hunk 1 above)
Comment (tfili) [inline Source/DataSources/dynamicGeometryGetBoundingSphere.js]: `GroundPrimitives` don't have a modelMatrix, they are always in the fixed frame.

File: Source/DataSources/RectangleGeometryUpdater.js (hunk 3 above)
Comment (tfili) [inline Source/DataSources/RectangleGeometryUpdater.js]: Not sure. I think it had something to do with changing an older version 6 months ago and then merging in current master. I don't recall changing it.

**Comment by pjcozzi** [comment]
I don't know exactly how all the pieces fit together, but is it too big of a change to separate `static`, `dynamic`, and `ground`, instead of masquerading `ground` as `dynamic`?  Deep down in the engine, they are basically separated like this.  We did the same in STK.

**Comment by pjcozzi** [comment]
Can we delete the `entity-clamp-to-ground` branch since it was merged into this?

**Comment by pjcozzi** [comment]
> 1. There are problems with both the camera tracking and selection indicator because it is positioned where it thinks the entity is, not where it is actually drawn

I think this is fairly hard since the shadow volume is currently huge to account for min/max height across all of Earth.  Once we have min/max height for just the bounding rectangle (or k-DOP or whatever), we can expose something much more precise.  This will be part of the vector tile work for 3D Tiles since it is also a significant rendering optimization.

Is there something else I am missing for camera support here?

**Comment by pjcozzi** [comment]
> 1. The GroundPrimitive can't currently be batched, so performance is terrible when there are a bunch of them on the screen. Just 900 rectangles drops my fps to 20 (code below).

We'll have the same problem with 3D Tiles, and we'll at least want to batch all the polygons in the tile of the same color.  As we discussed offline, this won't fly for when the color changes dynamically.

The best strategy for entities is to just wait to see what we do for 3D Tiles since the problem will be significantly worse there, and we will have not choice but to aggressively optimize.

Also, the other performance difference is that the fillrate for ground primitives is way more than regular primitives; this will be significantly improved with the local min/max height.

**Comment by mramato** [comment]
> Is there something else I am missing for camera support here?

I assume your comment on shadow volumes applies only to geometry. The problem also exists for billboards and labels (and will for points and models too when we add support).

**Comment by pjcozzi** [comment]
> 2. Geometry outlines can't be clamped to ground which will confuse users who will end up reporting it as a bug.
3. Geometry materials other than color doesn't work, which will also confuse users who will end up reporting it as a bug (but may not be as big of a deal as outlines).
4. Obviously lines aren't supported yet. I don't think using corridors as lines will work well at all, but I'm willing to entertain the idea and see where it ends up. (and they don't support materials/outline either)

For outlines and polylines, wait for 3D Tiles vector tiles - or hack around with corridors in the meantime.

If we can prioritize it, it would be nice to get broader material support soon (once we can compute texture coordinates, it should support all the same materials).  @bagnell how many days do you think this would take?  Including the fast paths for different sized polygons.

**Comment by pjcozzi** [comment]
>> Is there something else I am missing for camera support here?

> I assume your comment on shadow volumes applies only to geometry. The problem also exists for billboards and labels (and will for points and models too when we add support).

Ah, yes.  Those will be easier.

**Comment by pjcozzi** [comment]
Most of the comments here seem to be around missing features that we also need for 3D Tiles.  Once this PR stabilizes if there is a reasonable way to just get "polygons on terrain" into the Entity API - even with some quirks - I think it is an incremental improvement that many users will be happy with.

**Comment by mramato** [comment]
> I don't know exactly how all the pieces fit together, but is it too big of a change to separate static, dynamic, and ground, instead of masquerading ground as dynamic? Deep down in the engine, they are basically separated like this. We did the same in STK.

The only problem with this suggestion is that you still also have to handle dynamic ground primitives too.  Otherwise, yes, we can absolutely have a StaticGroundGeometryColorBatch class that handles ground geometry (first creating individual primitives and later batching them by color.

I have a pretty good understanding of what needs to change under the hood to dramatically improve both performance and cleanliness of the entity visualization layer, to better batch the new features and common requested use cases from the forum, but it's a largish effort and I unfortunately haven't had the time to work on it.

**Comment by mramato** [comment]
> even with some quirks - I think it is an incremental improvement that many users will be happy with.

I'm not sure I agree.  There are too many caveats and performance isn't there (as of today).  However, if we can fix a fraction of the above issues and add support for GeoJSON and KML, then I could be convinced otherwise.  I really want the Entity API to support terrain, but we need to keep the quality bar high.

**Comment by bagnell** [comment]
>If we can prioritize it, it would be nice to get broader material support soon (once we can compute texture coordinates, it should support all the same materials). @bagnell how many days do you think this would take? Including the fast paths for different sized polygons.

I started in a branch. There is code for computing the attributes/uniforms necessary to get the texture coordinates. It isn't validated and I haven't started generating the texture coordinates in the shader. I believe there are three shader paths. Probably a couple weeks worth of work.

For full material support, ground primitives would need normal, tangent and bitangent attributes as well which is only possible when the terrain is an ellipsoid.

**Comment by bagnell** [comment]
>Is there something else I am missing for camera support here?

The camera would abruptly change position from collision with the terrain when the LOD changes.

**Comment by tfili** [comment]
@mramato Can you take a look at this again? I believe you can ignore the changes to `GroundPrimitive.js` as Dan's changes will be in master before/if this is merged.

I've added a `StaticGroundGeometryColorBatch` along with a test in the `GeometryVisualizerSpec` where all the other batch tests were.

I hooked it up to `Corridor`,  `Ellipse`, `Polygon` and `Rectangle` and added tests for when they should be drawn on terrain for dynamic & static updaters.

Added `oneTimeWarning` class along with a test suite for cases like when we tell users that outlines are disabled on terrain.

File: Source/Core/oneTimeWarning.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/*global define*/
+define([
+           './defined',
```
Comment (mramato) [inline Source/Core/oneTimeWarning.js]: We keep the formatting of the `define` section consistent.  To fix this, just run `npm run sortRequires` and then submit the files it updates.

Comment (tfili) [inline Source/Core/oneTimeWarning.js]: Fixed.

**Comment by mramato** [comment]
Load the attached geojson file with terrain and you will crash.  This is probably a bug at the primitive level, but we should probably fix it before shipping this. CC @bagnell 

[countries.geojson.txt](https://github.com/AnalyticalGraphicsInc/cesium/files/101125/countries.geojson.txt)

**Comment by mramato** [comment]
We also need a warning when a polygon on terrain has a material other than color, correct?  Since that won't get drawn on terrain either.

**Comment by mramato** [comment]
Please merge in master, since @bagnell's changes and fixes are all in.  Unfortunately, the batching did not help performance at all. I verified that the problem is related to our huge bounding volume.

The code here mostly looks okay, but I didn't review too thoroughly yet.  The general approach is fine given the current architecture.  That being said, this is still a long way from coming into master, there are just too many issues.  Here's a list of what I think (at a minimum) still needs to happen.

- [ ] Camera `flyTo` of an entity on terrain.
- [ ] Camera tracking of an entity on terrain
- [ ] Selection indicator of an entity on terrain
- [ ] Point and Model clamping
- [ ] Detecting if `EXT_FRAG_DEPTH` is missing and fallback to non-ground visualization; otherwise this branch breaks IE, Safari, and various mobile devices.
- [ ] Performance

Even with the above, we'll still have a bunch of caveats left, but they might not be a deal breaker at that point.

**Comment by mramato** [comment]
I moved my todo list to the original summary.  Hopefully we can get back this soonish.

**Comment by vrittis** [comment]
I'm interested in this issue for work but as far as I understand there are some limitations at this moment:
most notably from what I'm looking into that entities only wrap terrain with color material, not image
@mramato Do we have a ETA on when this feature could be completed? Or does the current amount of work render it evaluation-proof?

**Comment by mramato** [comment]
1. Create a project
2. Save it.
3. Switch tot he Projects tab.
4. Globe goes away and nothing happens.  Check the console and there are errors.

**Comment by mramato** [comment]
@vrittis I can't provide any sort of official ETA, but I personally hope that terrain support for the Entity API and full material support for ground primitives are both done before the end of the year.  Our work on 3D Tiles will hopefully have some beneficial overlap here.  (And 3D Tiles themselves may help with a lot of the use cases that people are currently using the Entity API for).

**Comment by pjcozzi** [comment]
What is the minimum amount of work needed to get a first cut of this into master?  Is it the task list above, https://github.com/AnalyticalGraphicsInc/cesium/pull/3406#issue-126028253?

**Comment by mramato** [comment]
In my opinion, yes, but we might be able to get away without Point and Model clamping (though this is already a common request on the mailing list).  Performance will have to be looked at at the end using typical geojson/czml/kml files and we could punt initially on it.

To me, the real question is, do we have a plan in place to fix the underlying large shadow volume problem, because I believe that's the cause of many of these issues and definitely causes the performance problem.  From my understanding, it's a data problem, because we need to know min/max terrain height, right?

One idea would be to ship a JSON file with rectangles with min/max height for sections of the earth.  Large enough that it's not too much data, but small enough that it gives much better values than hardcoding Mount Everset.

**Comment by pjcozzi** [comment]
We're going to address the shadow volumes with 3D Tiles and, for, the general case, we might be able to get min/max from the loaded terrain.

I would not let this hold up this PR; so far users of the primitive API have soon that they would rather have the feature soon even with pre-optimized performance.  Then can opt in if you want.

**Comment by colek42** [comment]
This is probably the top feature we are looking to implement in the coming weeks with our application.  I'll give my vote for getting it into master and optimizing the performance later.

**Comment by mramato** [comment]
> I would not let this hold up this PR; so far users of the primitive API have soon that they would rather have the feature soon even with pre-optimized performance. Then can opt in if you want.

All users will have to opt-in, that was a forgone conclusion.  Even if performance wasn't an issue (and I think it is a big one) that checklist was the bare minimum we could accept to get this into master.   We will have to have a big fat disclaimer on the opt-in options listing the myriad of issues they will encounter if they turn the option on.

**Comment by mramato** [comment]
> This is probably the top feature we are looking to implement in the coming weeks with our application. I'll give my vote for getting it into master and optimizing the performance later.

@colek42 Can you explain some of the details of your use case?  Trust me, no one wants to get terrain support into the Entity API more than I do; but I also don't want a giant step back in quality, performance, and cross-browser compatibility.  I'm curious if your use case would actually work with this PR.  What kind of data are you loading (polygons, polylines, billboards, etc)?  How much of it is there?  Is it time-dynamic?  How big of an area, geographic, do you look at at a time?  Anything info you could provide would be useful.

**Comment by colek42** [comment]
We have sensor footprints that are described by four corners on the earth.  If we implement terrain these footprints will either be floating or hidden by terrain.  While we many have hundreds of these sensors on the globe at any time we  have some server side code, as well as a webworker, that ensures that only those footprints in the camera's current bounding box are in the entity collection array.  We are looking to implement terrain data in the coming weeks, without clamping for ground-based entities we do not have a clear development path.  We are confident that we can work around the performance issues as long as we can display a dozen or so polygons on the earth.

**Comment by mramato** [comment]
@colek42  Just to be clear, performance is the least of my worries and not the ultimate hold up for this PR.  If the Entity API doesn't have support when you start coding, for the simple use case you describe it would be trivial to use the Primitive API.  That being said, hopefully we can get this in by the release of 1.20 in April.

**Comment by colek42** [comment]
@mramato We are just happy we have an excellent open source product we can work on.  Keep doing what you guys are doing.  We have been very happy with our transition from GE thus far.

**Comment by pjcozzi** [comment]
@mramato so it sounds like you would be OK with this going into master if the task list above (minus performance, initially) is finished, which is:

> * Camera flyTo and zoomTo of an entity on terrain.
> * Camera tracking of an entity on terrain
> * Selection indicator of an entity on terrain
> * Detecting if EXT_FRAG_DEPTH is missing and fallback to non-ground visualization; otherwise this branch breaks IE, Safari, and various mobile devices.
> * Point and Model clamping (maybe, https://github.com/AnalyticalGraphicsInc/cesium/pull/3406#issuecomment-189450572)

Is this anything else that is a near-term priority so that we can consolidate branches?

**Comment by mramato** [comment]
> Is this anything else that is a near-term priority so that we can consolidate branches?

I think that list covers it.  As for the maybe next to Point and Model clamping, it it were up to me I would implement it, since everyone wants it and from my understanding the only reason we are holding off on it over possible performance concerns; but since it's ultimately is needed; I'm not sure that's a good reason to hold off.

**Comment by tfili** [comment]
Closing to move to move the branch out of my fork.
